File: 백준/Silver/1697. 숨바꼭질/숨바꼭질.py
from collections import deque

N, K = map(int, input().split())
visited = [False] * 100001


def bfs(v):
    if N == K:
        return 0
    q = deque([(v, 0)])
    visited[v] = True

    while q:
        cur, time = q.popleft()
        for i in [cur - 1, cur + 1, 2 * cur]:
            if i == K:
                return time + 1
            elif 0 <= i <= 100000 and not visited[i]:
                visited[i] = True
                q.append((i, time + 1))


print(bfs(N))

File: 백준/Silver/1697. 숨바꼭질/숨바꼭질.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;



public class Main {
    static boolean[] visited = new boolean[100001];
    static int N, K;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        K = Integer.parseInt(st.nextToken());
        System.out.println(bfs(N, K));

    }

    public static int bfs(int loc, int goal) {
        if (loc == goal) return 0;
        Queue<Node> q = new LinkedList<>();
        q.offer(new Node(loc, 0));
        visited[loc] = true;
        while (!q.isEmpty()) {
            Node cur = q.poll();
            int curLoc = cur.vertex;
            int curTime = cur.time;

            int[] nextMoves = {curLoc - 1, curLoc + 1, curLoc * 2};
            for (int nextLoc : nextMoves) {
                if (nextLoc == K) {
                    return curTime + 1;
                } else if (0 <= nextLoc && nextLoc <= 100000 && !visited[nextLoc]) {
                    visited[nextLoc] = true;
                    q.offer(new Node(nextLoc, curTime + 1));
                }
            }
        }
        return -1;
    }

    static class Node {
        int vertex;
        int time;

        Node(int vertex, int time) {
            this.vertex = vertex;
            this.time = time;
        }
    }
}